File: typescript-examples/computer-use-automation/api/anthropic-computer-use.ts
import z from "zod";
import { BrowserContext, Page } from "playwright";
import { samplingLoop } from "../lib/anthropic/sampling-loop";

interface Params {
  query: string;        // The task you want the AI to perform
  apiKey?: string;       // Your Anthropic API key
  model?: string;       // Model to use (default: 'claude-sonnet-4-20250514')
  thinkingBudget?: number;  // Token budget for thinking (default: 1024)
  maxTokens?: number;   // Max tokens per request (default: 4096)
}

export default async function handler(
  params: Params,
  page: Page,
  _: BrowserContext,
) {
  const {
    query,
    model = 'claude-sonnet-4-20250514',
    thinkingBudget = 1024,
    maxTokens = 4096,
  } = params;
  let { apiKey } = params;
  if (!apiKey) {
    apiKey = process.env.ANTHROPIC_API_KEY;
  }

  // Note: The Playwright page is already initialized and ready to use
  // Claude can use the go_to_url action to navigate directly without opening a browser

  if (!query) {
    throw new Error('Query is required');
  }

  if (!apiKey) {
    throw new Error('API key is required');
  }

  // Set viewport size to match the computer tool's display dimensions
  await page.setViewportSize({
    width: 1280,
    height: 720,
  });

  try {
    // Run the sampling loop
    const finalMessages = await samplingLoop({
      model,
      messages: [{
        role: 'user',
        content: query,
      }],
      apiKey,
      thinkingBudget,
      maxTokens,
      playwrightPage: page,
    });

    // Extract the final result from the messages
    if (finalMessages.length === 0) {
      throw new Error('No messages were generated during the sampling loop');
    }

    const lastMessage = finalMessages[finalMessages.length - 1];
    if (!lastMessage) {
      throw new Error('Failed to get the last message from the sampling loop');
    }

    const result = typeof lastMessage.content === 'string'
      ? lastMessage.content
      : lastMessage.content.map(block =>
        block.type === 'text' ? block.text : ''
      ).join('');

    return {
      result,
      messageCount: finalMessages.length,
    };
  } catch (error) {
    console.error('Error in sampling loop:', error);
    throw error;
  }
}

File: python-examples/computer-use-automation/api/anthropic-computer-use.py
from playwright.async_api import Page
from typing import TypedDict, Dict, NotRequired
from lib.anthropic.utils.loop import sampling_loop
import os


class Params(TypedDict):
    query: str  # The task you want the AI to perform


async def automation(page: Page, params: Params | None = None, **_kwargs):
    if not params or not params.get("query"):
        raise ValueError("Query is required, please provide a query in the params")
    
    # Get API key from environment
    api_key = os.getenv("ANTHROPIC_API_KEY")
    if not api_key:
        raise ValueError("ANTHROPIC_API_KEY environment variable is required")
    
    # Hardcoded model
    model = "claude-sonnet-4-20250514"
    
    # Set viewport size to match the computer tool's display dimensions
    await page.set_viewport_size({"width": 1280, "height": 720})
    
    final_messages = await sampling_loop(
        model=model,
        messages=[{
            "role": "user",
            "content": params["query"],
        }],
        api_key=str(api_key),
        thinking_budget=1024,
        playwright_page=page,
    )

    # Extract the final result
    if not final_messages:
        raise ValueError("No messages were generated during the sampling loop")

    last_message = final_messages[-1]
    if not last_message:
        raise ValueError("Failed to get the last message from the sampling loop")

    result = ""
    if isinstance(last_message.get("content"), str):
        result = last_message["content"]  # type: ignore[assignment]
    else:
        result = "".join(
            block["text"] for block in last_message["content"]  # type: ignore[index]
            if isinstance(block, Dict) and block.get("type") == "text"
        )

    return {"result": result}
